File: android/src/main/java/com/emi/plugins/admob/appopen/AppOpenExecutor.java
package com.emi.plugins.admob.appopen;

import android.app.Activity;
import androidx.annotation.NonNull;

import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.emi.plugins.admob.AdMobNextGenPlugin;
import com.emi.plugins.admob.ActionCallback;

import com.google.android.libraries.ads.mobile.sdk.appopen.AppOpenAd;
import com.google.android.libraries.ads.mobile.sdk.appopen.AppOpenAdEventCallback;
import com.google.android.libraries.ads.mobile.sdk.common.AdLoadCallback;
import com.google.android.libraries.ads.mobile.sdk.common.AdRequest;
import com.google.android.libraries.ads.mobile.sdk.common.AdValue;
import com.google.android.libraries.ads.mobile.sdk.common.FullScreenContentError;
import com.google.android.libraries.ads.mobile.sdk.common.LoadAdError;

import java.util.Date;

public class AppOpenExecutor {

    private final AdMobNextGenPlugin plugin;
    private AppOpenAd appOpenAd;
    private String currentAdUnitId = "";

    private boolean isLoadingAd = false;
    private boolean isShowingAd = false;
    private long loadTime = 0;

    private long lastRequestTime = 0;

    public AppOpenExecutor(AdMobNextGenPlugin plugin) {
        this.plugin = plugin;
    }

    private boolean wasLoadTimeLessThanNHoursAgo(long numHours) {
        long dateDifference = new Date().getTime() - loadTime;
        long numMilliSecondsPerHour = 3600000L;
        return dateDifference < (numMilliSecondsPerHour * numHours);
    }

    private boolean isAdAvailable() {
        return appOpenAd != null && wasLoadTimeLessThanNHoursAgo(4);
    }

    public void load(Activity activity, PluginCall call, final ActionCallback callback) {
        String adUnitId = call.getString("adUnitId");
        if (adUnitId == null || adUnitId.isEmpty()) {
            callback.onError("Ad Unit ID is required.");
            return;
        }

        Double retryOpt = call.getDouble("retryInterval");
        long minLoadInterval = (retryOpt != null) ? retryOpt.longValue() : 5000L;
        long currentTime = System.currentTimeMillis();

        if ((currentTime - lastRequestTime) < minLoadInterval) {
            callback.onError("Request too fast. Please wait " + minLoadInterval + "ms to prevent invalid traffic.");
            return;
        }

        if (isLoadingAd || isAdAvailable()) {
            callback.onError("App open ad is already loading or available.");
            return;
        }

        this.currentAdUnitId = adUnitId;
        this.lastRequestTime = currentTime;
        isLoadingAd = true;

        activity.runOnUiThread(() -> {
            AppOpenAd.load(
                    new AdRequest.Builder(adUnitId).build(),
                    new AdLoadCallback<AppOpenAd>() {
                        @Override
                        public void onAdLoaded(@NonNull AppOpenAd ad) {
                            appOpenAd = ad;
                            isLoadingAd = false;
                            loadTime = new Date().getTime(); 

                            JSObject ret = new JSObject();
                            ret.put("adUnitId", currentAdUnitId);
                            plugin.notifyPluginListeners("onAppOpenAdLoaded", ret);

                            callback.onSuccess();
                        }

                        @Override
                        public void onAdFailedToLoad(@NonNull LoadAdError adError) {
                            isLoadingAd = false;
                            appOpenAd = null;

                            JSObject ret = new JSObject();
                            ret.put("error", adError.getMessage());
                            plugin.notifyPluginListeners("onAppOpenAdFailedToLoad", ret);

                            callback.onError(adError.getMessage());
                        }
                    }
            );
        });
    }

    public void show(Activity activity, ActionCallback callback) {
        if (isShowingAd) {
            callback.onError("App open ad is already showing.");
            return;
        }

        if (!isAdAvailable()) {
            callback.onError("App open ad is not ready or has expired.");
            return;
        }

        activity.runOnUiThread(() -> {
            appOpenAd.setAdEventCallback(new AppOpenAdEventCallback() {
                @Override
                public void onAdShowedFullScreenContent() {
                    plugin.notifyPluginListeners("onAppOpenAdShowed", new JSObject());
                }

                @Override
                public void onAdDismissedFullScreenContent() {
                    appOpenAd = null; 
                    isShowingAd = false;
                    plugin.notifyPluginListeners("onAppOpenAdDismissed", new JSObject());
                }

                @Override
                public void onAdFailedToShowFullScreenContent(@NonNull FullScreenContentError error) {
                    appOpenAd = null; 
                    isShowingAd = false;

                    JSObject ret = new JSObject();
                    ret.put("error", error.getMessage());
                    plugin.notifyPluginListeners("onAppOpenAdFailedToShow", ret);
                }

                @Override
                public void onAdImpression() {
                    plugin.notifyPluginListeners("onAppOpenAdImpression", new JSObject());
                }

                @Override
                public void onAdClicked() {
                    plugin.notifyPluginListeners("onAppOpenAdClicked", new JSObject());
                }

                @Override
                public void onAdPaid(@NonNull AdValue value) {
                    JSObject ret = new JSObject();
                    ret.put("adUnitId", currentAdUnitId);
                    ret.put("valueMicros", value.getValueMicros());
                    ret.put("currencyCode", value.getCurrencyCode());
                    ret.put("precisionType", value.getPrecisionType().name());
                    plugin.notifyPluginListeners("onAppOpenAdPaid", ret);
                }
            });

            isShowingAd = true;
            appOpenAd.show(activity);
            callback.onSuccess();
        });
    }
}

File: ios/Sources/AdMobNextGenPlugin/appopen/AppOpenExecutor.swift
import Capacitor
import Foundation
import GoogleMobileAds

@objc public class AppOpenExecutor: NSObject, FullScreenContentDelegate {

    private weak var plugin: AdMobNextGenPlugin?

    private var appOpenAd: AppOpenAd?
    private var currentAdUnitId: String = ""

    private var isLoadingAd: Bool = false
    private var isShowingAd: Bool = false
    private var loadTime: Date?

    private var lastRequestTime: TimeInterval = 0

    init(plugin: AdMobNextGenPlugin) {
        self.plugin = plugin
        super.init()
    }

    private func wasLoadTimeLessThanNHoursAgo(numHours: Double) -> Bool {
        guard let loadTime = loadTime else { return false }
        let timeIntervalInSeconds = numHours * 3600.0
        return Date().timeIntervalSince(loadTime) < timeIntervalInSeconds
    }

    private func isAdAvailable() -> Bool {
        return appOpenAd != nil && wasLoadTimeLessThanNHoursAgo(numHours: 4.0)
    }

    @objc public func loadAd(_ call: CAPPluginCall) {
        guard let plugin = self.plugin else { return }

        if !plugin.isInitialized {
            call.reject(
                "Google Mobile Ads SDK has not been initialized. Please call initialize() first."
            )
            return
        }

        let adUnitId = call.getString("adUnitId", "")
        if adUnitId.isEmpty {
            call.reject("Ad Unit ID is required.")
            return
        }

        let retryOpt = call.getDouble("retryInterval")
        let minLoadInterval: TimeInterval =
            (retryOpt != nil) ? (retryOpt! / 1000.0) : 5.0  
        let currentTime = Date().timeIntervalSince1970

        if (currentTime - lastRequestTime) < minLoadInterval {
            call.reject(
                "Request too fast. Please wait \(minLoadInterval * 1000)ms to prevent invalid traffic."
            )
            return
        }

        if isLoadingAd || isAdAvailable() {
            call.reject("App open ad is already loading or available.")
            return
        }

        self.currentAdUnitId = adUnitId
        self.lastRequestTime = currentTime
        self.isLoadingAd = true

        let request = Request()

        Task { @MainActor in
            do {
                self.appOpenAd = try await AppOpenAd.load(
                    with: adUnitId,
                    request: request
                )
                self.appOpenAd?.fullScreenContentDelegate = self

                self.appOpenAd?.paidEventHandler = { [weak self] value in
                    guard let self = self else { return }
                    var ret = JSObject()
                    ret["adUnitId"] = self.currentAdUnitId
                    ret["valueMicros"] = value.value.stringValue  
                    ret["currencyCode"] = value.currencyCode

                    ret["precisionType"] = value.precision.rawValue
                    self.plugin?.notifyListeners("onAppOpenAdPaid", data: ret)
                }

                self.isLoadingAd = false
                self.loadTime = Date()

                var ret = JSObject()
                ret["adUnitId"] = self.currentAdUnitId
                plugin.notifyListeners("onAppOpenAdLoaded", data: ret)

                call.resolve()
            } catch {
                self.isLoadingAd = false
                self.appOpenAd = nil

                var ret = JSObject()
                ret["error"] = error.localizedDescription
                plugin.notifyListeners("onAppOpenAdFailedToLoad", data: ret)

                call.reject(error.localizedDescription)
            }
        }
    }

    @objc public func showAd(_ call: CAPPluginCall) {
        guard let plugin = self.plugin else { return }

        if isShowingAd {
            call.reject("App open ad is already showing.")
            return
        }

        if !isAdAvailable() {
            call.reject("App open ad is not ready or has expired.")
            return
        }

        DispatchQueue.main.async {
            guard let viewController = plugin.bridge?.viewController else {
                call.reject("Root ViewController not found.")
                return
            }

            if let ad = self.appOpenAd {
                self.isShowingAd = true
                ad.present(from: viewController)
                call.resolve()
            }
        }
    }

    public func adWillPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
        plugin?.notifyListeners("onAppOpenAdShowed", data: [:])
    }

    public func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
        appOpenAd = nil  
        isShowingAd = false
        plugin?.notifyListeners("onAppOpenAdDismissed", data: [:])
    }

    public func ad(
        _ ad: FullScreenPresentingAd,
        didFailToPresentFullScreenContentWithError error: Error
    ) {
        appOpenAd = nil  
        isShowingAd = false

        var ret = JSObject()
        ret["error"] = error.localizedDescription
        plugin?.notifyListeners("onAppOpenAdFailedToShow", data: ret)
    }

    public func adDidRecordImpression(_ ad: FullScreenPresentingAd) {
        plugin?.notifyListeners("onAppOpenAdImpression", data: [:])
    }

    public func adDidRecordClick(_ ad: FullScreenPresentingAd) {
        plugin?.notifyListeners("onAppOpenAdClicked", data: [:])
    }

}
